include/cuco/static_set_ref.cuh

```diff
@@ -0,0 +1,155 @@
+  using storage_ref_type    = StorageRef;                      ///< Type of slot storage ref
+  using window_type = typename storage_ref_type::window_type;  ///< Probing scheme element type
+  using value_type  = typename storage_ref_type::value_type;   ///< Probing scheme element type
+  using size_type   = typename storage_ref_type::size_type;    ///< Probing scheme size type
+  using key_equal =
+    detail::equal_wrapper<value_type, KeyEqual>;  ///< Type of key equality binary callable
+
+  static constexpr int cg_size = probing_scheme_type::cg_size;  ///< Cooperative group size
+  static constexpr int window_size =
+    storage_ref_type::window_size;                    ///< Number of elements handled per window
+  static constexpr cuda::thread_scope scope = Scope;  ///< Thread scope
+
+  // TODO default ctor?
+
+  /**
+   * @brief Constructs static_set_ref.
+   *
+   * @param empty_key_sentinel Sentinel indicating empty key
+   * @param predicate Key equality binary callable
+   * @param probing_scheme Probing scheme
+   * @param storage_ref Non-owning ref of slot storage
+   */
+  __host__ __device__ static_set_ref(cuco::sentinel::empty_key<Key> empty_key_sentinel,
+                                     KeyEqual const& predicate,
+                                     ProbingScheme const& probing_scheme,
+                                     StorageRef storage_ref) noexcept
+    : empty_key_sentinel_{empty_key_sentinel},
+      predicate_{empty_key_sentinel_.value, predicate},
+      probing_scheme_{probing_scheme},
+      storage_ref_{storage_ref}
+  {
+  }
+
+  /**
+   * @brief Gets the maximum number of elements the hash map can hold.
+   *
+   * @return The maximum number of elements the hash map can hold
+   */
+  [[nodiscard]] __host__ __device__ inline size_type capacity() const noexcept
+  {
+    return storage_ref_.capacity();
+  }
+
+  /**
+   * @brief Gets the sentinel value used to represent an empty key slot.
+   *
+   * @return The sentinel value used to represent an empty key slot
+   */
+  [[nodiscard]] __host__ __device__ inline key_type empty_key_sentinel() const noexcept
+  {
+    return empty_key_sentinel_;
+  }
+
+  /**
+   * @brief Create a reference with functions.
+   *
+   * @tparam NewFunctions List of `cuco::function::*` types
+   */
+  template <typename... NewFunctions>
+  using make_with_functions =
+    static_set_ref<Key,
+                   Scope,
+                   KeyEqual,
+                   ProbingScheme,
+                   StorageRef,
+                   NewFunctions...>;  ///< Type alias for the current ref type with a new set of
+                                      ///< functions
+
+  /**
+   * @brief Create a reference with new functions from the current object.
+   *
+   * @tparam NewFunctions List of `cuco::function::*` types
+   *
+   * @return copy of `*this` with `newFunctions`
+   */
+  template <typename... NewFunctions>
+  [[nodiscard]] __host__ __device__ auto with_functions() const
```

Comment:
We'll want to provide solid documentation about the restriction against concurrent use of references to the same storage with different functions.

Reply:
Sure, otherwise this is a proper footgun. I would even suggest to add the op/op compatibility matrix that I'm working on to the documentation at some point. Could even be part of some sort of tuning guide.

Reply:
I hope [03adc20](https://github.com/NVIDIA/cuCollections/pull/246/commits/03adc20c51f47bcf31a96ed91182daabc1313f52) sound dramatic enough :D